File: 2021/06/30/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	var findLCA func(*TreeNode, *TreeNode, *TreeNode) *TreeNode

	findLCA = func(root, p, q *TreeNode) *TreeNode {
		if root == nil || root == p || root == q {
			return root
		}

		l := findLCA(root.Left, p, q)
		r := findLCA(root.Right, p, q)

		if l != nil && r != nil {
			return root
		}
		if l != nil {
			return l
		}
		return r
	}
	return findLCA(root, p, q)
}

File: 2021/06/30/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def findLCA(root: 'TreeNode', p: 'TreeNode', q: 'TreeNode'):
            if root == None or root == p or root == q:
                return root
            
            l = findLCA(root.left, p, q)
            r = findLCA(root.right, p, q)

            if l and r:
                return root
            
            return l if l is not None else r
        
        return findLCA(root, p, q)